text/0072-online-bulk-load-for-rawkv.md

```diff
@@ -0,0 +1,103 @@
+# RFC: Online Bulk Load for RawKV
+
+- RFC PR: https://github.com/tikv/rfcs/pull/72
+- Tracking Issue: https://github.com/tikv/tikv/issues/10563
+- [Proposal: TiKV Online Bulk Load (Chinese)](https://docs.google.com/document/d/1JggGwmOIMqinz_Q7QkIDHzwS--K_BgJopaxrB8Pzpaw/edit#)
+
+## Summary
+
+Let RawKV support Online Bulk Load.
+
+## Motivation
+
+In the use use of `Feature Store`, every day machine learning jobs will train a
+large amount of data which need to load to TiKV for online serving.
+
+![Feature Store](../media/feature-store-tikv.png)
+
+In order to satisfy this scenario, TiKV needs to provide `online bulk load`,
+which should meet the following requirements:
+
+1. the loading process should be as fast as possible,
+2. the impact to the online serving (P99 latency) should be under control,
+3. the loading framework should be able to process Big Data (500GB ~ 5TB).
+
+## Detailed design
+
+The solution proposed in this RFC is as follows:
+
+1. use distributed computing framework `Spark` to read the source files, convert
+and encode the trained data into Key Value format,
+2. sample the Key Value pairs, calculate the region split points, and call
+`split region and scatter` API
+3. repartition the Key Value pairs RDD according to the new Regions, sort inside each partition,
+use the [Import SST API](https://github.com/pingcap/kvproto/blob/release-5.0/proto/import_sstpb.proto#L34)
+to concurrently send the Key-Value pairs to the TiKV server and generate SST
+files on the TiKV server,
+4. use the [Ingest API](https://github.com/pingcap/kvproto/blob/release-5.0/proto/import_sstpb.proto#L53)
+to ingest the generated SST files into TiKV.
+
+![Bulk Load](../media/bulk-load.png)
+
+### Proto
+
+```
+service ImportSST {
+  rpc RawWrite(stream RawWriteRequest) returns (RawWriteResponse) {}
+}
+
+message RawWriteBatch {
+    uint64 ttl = 1;
+    repeated Pair pairs = 2;
+}
+
+message RawWriteRequest {
+    oneof chunk {
+        SSTMeta meta = 1;
+        RawWriteBatch batch = 2;
+    }
+}
+
+message RawWriteResponse {
+    Error error = 1;
+    repeated SSTMeta metas = 2;
+}
+```
+
+### Rate Limit
```

Comment:
Perhaps the rate limit is out of topic for this RFC because it's one of the features of the io rate limiter but not introduced by the online bulk load.

Reply:
Users should set the following configurations to produce a good trade-off between ingesting speed and impact on the online service

- config storage.io-rate-limit.max-bytes-per-sec
- config storage.io-rate-limit.import-priority = "low"

Reply:
why is `storage.io-rate-limit` set to off by default? I thought it should default to `on` to mitigate the performance jitter since TiDB 5.0 :joy:

Reply:
Cause users should firstly get the value `max-bytes-per-sec` on their own hardware by running `fio`.

Reply:
It's better that TiKV it self can test and calculate the value `max-bytes-per-sec`.